privaterelay/cleaners.py

```diff
@@ -1,101 +1,45 @@
-        if not self._cleanup_data:
-            raise ValueError(
-                "self._cleanup_data must have a value when calling cleanup_data."
-            )
-        return self._cleanup_data
-
-    def issues(self) -> int:
-        """Return the number of detected data issues."""
-        return self.counts["summary"]["needs_cleaning"]
-
-    def _get_counts_and_data(self) -> tuple[Counts, CleanupData]:
-        """Return a dictionary of counts and cleanup data."""
-        raise NotImplementedError("_get_counts_and_data() not implemented")
-
-    def _clean(self) -> int:
-        """
-        Clean the detected items.
-
-        Returns the number of cleaned items. Implementors can add detailed
-        counts to self._counts as needed.
-        """
-        raise NotImplementedError("_clean() not implemented")
-
-    def clean(self) -> int:
-        """Clean the detected items, and update counts["summary"]"""
-        summary = self.counts["summary"]
-        if not self._cleaned:
-            summary["cleaned"] = self._clean()
-            self._cleaned = True
-        return summary["cleaned"]
-
-    def markdown_report(self) -> str:
-        """Return Markdown-formatted report of issues found and (maybe) fixed."""
-        raise NotImplementedError("markdown_report() not implemented")
-
-    @staticmethod
-    def _as_percent(part: int, whole: int) -> str:
-        """Return value followed by percent of whole, like '5 ( 30.0%)'"""
-        if not whole > 0:
-            raise ValueError("whole must be greater than 0 when calling _as_percent")
-        len_whole = len(str(whole))
-        return f"{part:{len_whole}d} ({part / whole:6.1%})"
-
-
-class CleanerTask(DataIssueTask):
-    """Base class for tasks that can clean up detected issues."""
-
-    can_clean = True
-
-
-class DetectorTask(DataIssueTask):
-    """Base class for tasks that cannot clean up detected issues."""
-
-    can_clean = False
-
-    def _clean(self) -> int:
-        """DetectorTask can't clean any detected issues."""
-        return 0
+"""Tasks that detect and fix data issues in privaterelay app or 3rd party apps."""
+
+from django.contrib.auth.models import User
+from django.db.models import Q, QuerySet
+
+from allauth.socialaccount.models import SocialAccount
+
+from .cleaner_task import CleanerTask, DataBisectSpec, DataModelSpec
+
+
+class MissingEmailCleaner(CleanerTask):
+    slug = "missing-email"
+    title = "Ensure all users have an email"
+    check_description = (
+        "When User.email is empty, we are unable to forward email to the Relay user."
+        " We can get the email from the FxA profile if available."
+    )
+
+    data_specification = [
+        # Report on how many users do not have an email
+        DataModelSpec(
+            model=User,
+            subdivisions=[
+                DataBisectSpec("email", ~Q(email__exact="")),
+                DataBisectSpec("!email.fxa", Q(socialaccount__provider="fxa")),
```

Comment:
The hard-coded `fxa` here means that the application must be set up with "fxa" as the name of the `SocialAccount` `SocialApp` name or this code will fail to find the user's FXA, which caused my error with the spot-check.

suggestion (blocking): either change this value to check for something like `Q(socialaccount__provider=SocialAccount.SocialApp.get(provider="fxa").name)` (not exactly sure of the syntax), or let's update the README directions to always use "fxa" for the provider id+name.

Reply:
Can you tell me more about the bug you were seeing? The query `Q(socialaccout_provider="fxa")` means what you are suggesting - the `SocialAccount` where `provider=fxa`, not the name.

Reply:
Ah, I might be getting `SocialAccount` and `SocialApplication` mixed up - I'll dig in.

Reply:
No, I'm not. `SocialApp` has a provider that is `fxa` for Firefox Accounts, and `SocialAccount` should have provider set to `fxa` as well. What is `SocialAccount.provider` for the test user that wasn't working?

Reply:
`SocialAccount.provider` for the user was `accounts.stage.mozaws.net` when it wasn't working, which seems to be copied from `SocialApplication` at creation-time ...

### Test 1: `SocialApplication` "fxa"
<img width="521" alt="image" src="https://github.com/mozilla/fx-private-relay/assets/71928/6663e51b-1d21-4061-9a2c-b302fa4c0ba6">

1. Sign-up new user
2. Check http://127.0.0.1:8000/admin/socialaccount/socialaccount/
   * [ ] provider is "fxa"
3. Delete the `User.email` value
4. Run `python manage.py cleanup_data -v 2 --missing-email --clean`
   * [ ] Fixes the issue! (`Cleaned: 1 (100.0%)`)

### Test 2: `SocialApplication` "accounts.stage.mozaws.net"
<img width="527" alt="image" src="https://github.com/mozilla/fx-private-relay/assets/71928/dfee40ed-85a5-4436-985c-e72f2e61d097">

1. Change SocialApplication "Provider ID" and "Name" value to `accounts.stage.mozaws.net`
2. Sign-up new user
3. Check http://127.0.0.1:8000/admin/socialaccount/socialaccount/
   * [ ] provider is "accounts.stage.mozaws.net"
4. Delete the `User.email` value
5. Run `python manage.py cleanup_data -v 2 --missing-email --clean`
   * [ ] Doesn't find the FXA! (`Has FxA: 0 (  0.0%)`)

Reply:
Ah, thanks. Mine is looks like this:

<img width="482" alt="Screenshot 2024-06-13 at 12 58 02 PM" src="https://github.com/mozilla/fx-private-relay/assets/286017/927f2253-e30d-4cbb-8b81-655a7fede6f7">

The optional `provider_id` was added in [0.55.0](https://docs.allauth.org/en/latest/release-notes/recent.html#id47). I'll dig in and see what that is for.

Reply:
TL;DR - you should not be setting `provider_id`, leave it blank like it is in dev, stage, and prod. However, I'll update the code to handle it.

The documentation is not very helpful. There's specific provider docs, and this doc for `allauth.socialaccount.adapter.DefaultSocialAccountAdapter.get_provider()`:

> Looks up a `provider`, supporting subproviders by looking up by `provider_id`.

https://docs.allauth.org/en/latest/socialaccount/adapter.html#allauth.socialaccount.adapter.DefaultSocialAccountAdapter.get_provider

The code is a little more helpful:

> For providers that support subproviders, such as OpenID Connect and SAML, this ID identifies that instance. SocialAccount's originating from app will have their `provider` field set to the `provider_id` if available, else `provider`.

https://github.com/pennersr/django-allauth/blob/9be5ef00e8d1ea697baf046c00957f259133aae5/allauth/socialaccount/models.py#L47-L55

When a provider is providing OpenID Connect or SAML, then `django-allauth` has been removing the provider and using the generic provider:

* https://docs.allauth.org/en/latest/socialaccount/providers/authelia.html
* https://docs.allauth.org/en/latest/socialaccount/providers/cern.html
* https://docs.allauth.org/en/latest/socialaccount/providers/openid_connect.html
* https://docs.allauth.org/en/latest/socialaccount/providers/saml.html

Reply:
The latest commit fixes the `provider_id` issue, with tests.